climada/util/yearsets.py

```diff
@@ -0,0 +1,244 @@
+
+LOGGER = logging.getLogger(__name__)
+
+def impact_yearset(event_impacts, sampled_years=None, sampling_vect=None, correction_fac=True):
+
+    """Create an annual_impacts object containing a probabilistic impact for each year
+      in the sampled_years list (or for a list of sampled_years generated with the length
+                                 of given sampled_years)
+      by sampling events from the existing input event_impacts with a Poisson distribution
+      centered around n_events per year (n_events = sum(event_impacts.frequency)).
+
+    Parameters:
+      event_impacts : impact object
+          impact object containing impacts per event
+    Optional parameters:
+        sampled_years : int or list
+            Either an integer specifying the amount of years to
+            be sampled or a list of years that shall be covered by the resulting annual_impacts
+            default: a 1000 year-long list starting in the year 0001
+        sampling_vect : dict
+            The sampling vector specifying which events contained in the events_impacts are
+            selected in the newly created annual_impacts and the number of events per year
+            in the annual_impacts, therefore the sampling_vect contains two arrays:
+                selected_events (array): sampled events (len: total amount of sampled events)
+                events_per_year (array): events per sampled year
+            The sampling_vector needs to be obtained in a first call,
+            i.e. [annual_impacts, sampling_vect] = climada_yearsets.impact_yearset(...)
+            and can then be provided in subsequent calls(s) to obtain the exact same sampling
+            (also for a different event_impacts object)
+        correction_fac : boolean
+            If True the resulting annual_impacts are scaled in so that the expected annual
+            impact (eai) of the annual_impacts = eai of the events_impacts
+
+    Returns:
+      annual_impacts : impact object
+          annual impacts for all sampled_years
+      sampling_vect : dict
+          the sampling vector containing two arrays:
+              selected_events (array) : sampled events (len: total amount of sampled events)
+              events_per_year (array) : events per sampled year
+          Can be used to re-create the exact same annual_impacts yearset
+      """
+
+
+    if not sampled_years and not sampling_vect:
+        sampled_years = np.arange(1, 1000+1).tolist()
+    elif not sampled_years:
+        sampled_years = np.arange(1, len(sampling_vect['selected_events'])+1).tolist()
+    elif isinstance(sampled_years, int):
+        sampled_years = np.arange(1, sampled_years+1).tolist()
+
+    if sampling_vect:
+        if len(sampled_years) != len(sampling_vect['events_per_year']):
+            raise ValueError("The number of sampled_years and the length of the list of "
+                             "events_per_year in the sampling vector must be equal.")
```

Comment:
Would it not be more logical to infer `sampled_years` from `sampling_vect` ? I do not really see a possiblity to define these two separately. Then you do not need the error message, just a warning maybe that `sampled_years` was ignored.

Reply:
Yes, it can (line 69) - I'll add this case there